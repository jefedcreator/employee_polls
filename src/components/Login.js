import React, { useState } from 'react'
import { connect } from 'react-redux'
import { useNavigate } from 'react-router-dom'
import { handleLogin } from '../actions/authedUser'


const Login = ({dispatch}) => {
    const [details, setDetails] = useState({
        username:"",
        password:""
    })

    const navigate = useNavigate()

    const handleChange = (e) =>{
        const {name, value} = e.target;
        setDetails({...details,[name]:value})
    }

    const handleClick = (e) =>{
        e.preventDefault()
        dispatch(handleLogin(details.username,details.password))
        navigate('/questions')
        setDetails({username:"",password:""})
    }


  return (
    <div className='flex flex-col items-center justify-center h-screen'>
        <h1 className='h-1/3 w-4/5 font-bold'>Employee Polls</h1>
        <div className='h-2/3'>
            <form className='flex flex-col h-2/3 justify-between'>
                <input type="text" placeholder='employee id' value={details.username} name='username' onChange={handleChange} className='p-2 outline-none border-2 border-indigo-500' data-testid='username-input'/>
                <input type="text" placeholder='password' value={details.password} name='password' onChange={handleChange} className='p-2 outline-none border-2 border-indigo-500' data-testid='password-input'/>
                <button className='bg-indigo-500 px-1 py-1 text-slate-200' onClick={handleClick} data-testid="submit-button">Login</button>
            </form>
        </div>
    </div>
  )
}

const mapStateToProps = ({ authedUser, users }) => ({
    authedUser,
    users
})

export default connect(mapStateToProps)(Login)