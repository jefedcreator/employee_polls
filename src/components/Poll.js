import React, { useState } from 'react'
import { connect } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { handleAddQuestion } from '../actions/questions';


const Poll = ({dispatch, authedUser}) => {
    const [poll, setPoll] = useState({
        optionOneText: "",
        optionTwoText: "",
    });

    const navigate = useNavigate()

    const handlePoll = (e) =>{
        const {name, value} = e.target;
        setPoll({...poll, [name]: value})
    }
    console.log("Poll options:", poll);

    const handleSubmit = (e) =>{
        e.preventDefault();
        dispatch(handleAddQuestion({
            optionOneText: poll.optionOneText,
            optionTwoText: poll.optionTwoText,
            author:authedUser
        }))
        setPoll({optionOneText:"",optionTwoText:""})
        navigate('/home')
    }

  return (
    <div className='h-full flex flex-col'>
        <h2 className='font-bold'>Would you Rather?</h2>
        <div className='h-4/5 flex flex-col justify-around items-center'>
            <p>Create your own poll</p>
            <form className='h-1/2 flex flex-col justify-between'>
                <input name='optionOneText' type="text" placeholder='first question' value={poll.optionOneText} onChange={handlePoll} className='outline-indigo-500 p-3 border-2 border-indigo-500'/>
                <input name='optionTwoText' type="text" placeholder='second question' value={poll.optionTwoText} onChange={handlePoll} className='outline-indigo-500 p-3 border-2 border-indigo-500'/>
                <button onClick={handleSubmit} className='bg-indigo-500 px-3 py-2 text-slate-200 cursor-pointer'>
                    Create
                </button>
            </form>
        </div>
    </div>
  )
}

const mapStateToProps = ({ authedUser }) => ({
    authedUser
})

export default connect(mapStateToProps)(Poll)