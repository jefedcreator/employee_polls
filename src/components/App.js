import { Fragment, useEffect } from 'react';
import { connect } from 'react-redux';
import { handleInitialData } from '../actions/shared';
import Home from './Home'
import Login from './Login';
import { LoadingBar } from 'react-redux-loading-bar';


function App({dispatch, authedUser}) {
  
  useEffect(()=>{
    dispatch(handleInitialData())
  },[]);

  return (
    <Fragment>
      <LoadingBar/>
      <div className="container px-10 py-5 h-screen">
        {
          authedUser == null ? 
          <Login/> 
          :
           <Home/>
        }

      </div>
    </Fragment>
  );
}

const mapStateToProps = ({ questions,users, authedUser }) => ({
  questions: Object.keys(questions).sort((a,b) => questions[b].timestamp - questions[a].timestamp),
  users: Object.keys(users).sort((a,b) => users[b].questions.length - users[a].questions.length),
  usersObj: users,
  authedUser,
  questionsObj: questions,
  loading: authedUser === null || undefined
})


export default connect(mapStateToProps)(App);
