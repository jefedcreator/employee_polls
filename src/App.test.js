import { render, screen, fireEvent } from '@testing-library/react'
import App from './components/App';
import {Provider} from 'react-redux';
import { MemoryRouter } from 'react-router';
import store from './store'
import { _saveQuestion, _saveQuestionAnswer } from "./utils/_DATA";
import Login from './components/Login';
import Poll from './components/Poll';

describe('saveQuestion', () => { 
  it('should resolve formatted questions', async () =>{
      let question = {
          optionOneText: "random",
          optionTwoText: "not random",
          author:"me"
      }

      let formattedQuestion = await _saveQuestion(question)
      
      expect(formattedQuestion.optionOne.text).toEqual(question.optionOneText)
      expect(formattedQuestion.optionTwo.text).toEqual(question.optionTwoText)
      expect(formattedQuestion.author).toEqual(question.author)

  })

  it('should reject invalid formatted questions', async () =>{
      let question = {
          optionOneText: undefined,
          optionTwoText: "not random",
          author:"me"
      }
      
      await expect(_saveQuestion(question)).rejects.toEqual("Please provide optionOneText, optionTwoText, and author")

  })
})

describe('saveQuestionAnswer',() =>{
  it('should resolve formatted question and answer', async ()=>{
    let answer = {
      authedUser:"zoshikanlu",
      id: "xj352vofupe1dqz9emx13r",
      voteId: 1,
    }

    let { authedUser, id, voteId } = answer

    let formatQuestionAnswer = await _saveQuestionAnswer(authedUser, id, voteId)
    expect(formatQuestionAnswer).toEqual(true)
  })

  it('should reject invalid formatted question and answer', async ()=>{
    let answer = {
      authedUser:undefined,
      id: undefined,
      voteId: 1,
    }

    let { authedUser, id, voteId } = answer

    await expect(_saveQuestionAnswer(authedUser, id, voteId)).rejects.toEqual("Please provide authedUser, qid, and answer")
  })
})

describe('Login', () => { 
  it('should test login component', () =>{
    let component = render(
    <MemoryRouter>
      <Provider store={store}>
        <Login/>
      </Provider>
    </MemoryRouter>
    )
    expect(component).toMatchSnapshot()
  })
  it('should have all expected fields', () =>{
    let component = render(
    <MemoryRouter>
      <Provider store={store}>
        <Login/>
      </Provider>
    </MemoryRouter>
    )
    
    let usernameInput = component.getByTestId('username-input')
    let passwordInput = component.getByTestId('password-input')

    expect(usernameInput).toBeInTheDocument()
    expect(passwordInput).toBeInTheDocument()
  })
  it('should test login upon click event', () =>{
    let component = render(
    <MemoryRouter>
      <Provider store={store}>
        <Login/>
      </Provider>
    </MemoryRouter>
    )
    let usernameInput = component.getByTestId('username-input')
    let passwordInput = component.getByTestId('password-input')
    let loginButton = component.getByTestId('submit-button')
   
    fireEvent.change(usernameInput, {target : { value : 'sarahedo', name : 'username'}})
    fireEvent.change(passwordInput, {target : { value : 'password123', name : 'password'}})
    fireEvent.click(loginButton);
  })
})

describe('Create Poll', () => { 
  it('should test crete poll component', () =>{
    let component = render(
    <MemoryRouter>
      <Provider store={store}>
        <Poll/>
      </Provider>
    </MemoryRouter>
    )
    expect(component).toMatchSnapshot()
  })
  it('should have all expected fields', () =>{
    let component = render(
    <MemoryRouter>
      <Provider store={store}>
        <Poll/>
      </Provider>
    </MemoryRouter>
    )
    
    let firstOption = component.getByTestId('firstoption-input')
    let secondOption = component.getByTestId('secondoption-input')

    expect(firstOption).toBeInTheDocument()
    expect(secondOption).toBeInTheDocument()
  })
  it('should create poll upon click event', () =>{
    let component = render(
    <MemoryRouter>
      <Provider store={store}>
        <Poll/>
      </Provider>
    </MemoryRouter>
    )
    let firstOption = component.getByTestId('firstoption-input')
    let secondOption = component.getByTestId('secondoption-input')
    let loginButton = component.getByTestId('submit-button')
   
    fireEvent.change(firstOption, {target : { value : 'option one'}})
    fireEvent.change(secondOption, {target : { value : 'option two'}})
    fireEvent.click(loginButton);
  })
})
