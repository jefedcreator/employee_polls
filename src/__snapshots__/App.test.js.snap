// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Poll should test crete poll component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-full flex flex-col"
      >
        <h2
          class="font-bold"
        >
          Would you Rather?
        </h2>
        <div
          class="h-4/5 flex flex-col justify-around items-center"
        >
          <p>
            Create your own poll
          </p>
          <form
            class="h-1/2 flex flex-col justify-between"
          >
            <input
              class="outline-indigo-500 p-3 border-2 border-indigo-500"
              data-testid="firstoption-input"
              name="optionOneText"
              placeholder="first question"
              type="text"
              value=""
            />
            <input
              class="outline-indigo-500 p-3 border-2 border-indigo-500"
              data-testid="secondoption-input"
              name="optionTwoText"
              placeholder="second question"
              type="text"
              value=""
            />
            <button
              class="bg-indigo-500 px-3 py-2 text-slate-200 cursor-pointer"
              data-testid="submit-button"
            >
              Create
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-full flex flex-col"
    >
      <h2
        class="font-bold"
      >
        Would you Rather?
      </h2>
      <div
        class="h-4/5 flex flex-col justify-around items-center"
      >
        <p>
          Create your own poll
        </p>
        <form
          class="h-1/2 flex flex-col justify-between"
        >
          <input
            class="outline-indigo-500 p-3 border-2 border-indigo-500"
            data-testid="firstoption-input"
            name="optionOneText"
            placeholder="first question"
            type="text"
            value=""
          />
          <input
            class="outline-indigo-500 p-3 border-2 border-indigo-500"
            data-testid="secondoption-input"
            name="optionTwoText"
            placeholder="second question"
            type="text"
            value=""
          />
          <button
            class="bg-indigo-500 px-3 py-2 text-slate-200 cursor-pointer"
            data-testid="submit-button"
          >
            Create
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login should test login component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col items-center justify-center h-screen"
      >
        <h1
          class="h-1/3 w-4/5 font-bold"
        >
          Employee Polls
        </h1>
        <div
          class="h-2/3"
        >
          <form
            class="flex flex-col h-2/3 justify-between"
          >
            <input
              class="p-2 outline-none border-2 border-indigo-500"
              data-testid="username-input"
              name="username"
              placeholder="employee id"
              type="text"
              value=""
            />
            <input
              class="p-2 outline-none border-2 border-indigo-500"
              data-testid="password-input"
              name="password"
              placeholder="password"
              type="text"
              value=""
            />
            <button
              class="bg-indigo-500 px-1 py-1 text-slate-200"
              data-testid="submit-button"
            >
              Login
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col items-center justify-center h-screen"
    >
      <h1
        class="h-1/3 w-4/5 font-bold"
      >
        Employee Polls
      </h1>
      <div
        class="h-2/3"
      >
        <form
          class="flex flex-col h-2/3 justify-between"
        >
          <input
            class="p-2 outline-none border-2 border-indigo-500"
            data-testid="username-input"
            name="username"
            placeholder="employee id"
            type="text"
            value=""
          />
          <input
            class="p-2 outline-none border-2 border-indigo-500"
            data-testid="password-input"
            name="password"
            placeholder="password"
            type="text"
            value=""
          />
          <button
            class="bg-indigo-500 px-1 py-1 text-slate-200"
            data-testid="submit-button"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
